src/python/yapykaldi/asr.py

```diff
@@ -0,0 +1,161 @@
+from __future__ import (print_function, division, absolute_import, unicode_literals)
+from builtins import *
+import os
+import datetime
+import time
+import struct
+import logging
+import multiprocessing
+import signal
+import wave
+import errno
+from string import Template
+import numpy as np
+import pyaudio
+from .nnet3 import KaldiNNet3OnlineDecoder, KaldiNNet3OnlineModel
+
+
+logging.basicConfig(level=logging.INFO,
+                    format='(%(processName)-9s) %(message)s',)
+
+
+def makedir_exist_ok(dirpath):
+    """
+    Python2 support for os.makedirs(.., exist_ok=True)
+    """
+    try:
+        os.makedirs(dirpath)
+    except OSError as e:
+        if e.errno == errno.EEXIST:
+            pass
+        else:
+            raise
+
+
+class Asr(object):
+    """API for ASR"""
+    def __init__(self, model_dir, model_type, output_dir, format=pyaudio.paInt16, channels=1, rate=16000, chunk=1024,
+                 timeout=2, wav_out_fmt="$date-$time"):
+        """
+        :param model_dir: Path to model directory
+        :param model_type: Type of ASR model 'nnet3' or 'hmm'
+        :param output_dir: Path to the directory where the recorded audio files will be written
+        :param format: (default pyaudio.paInt16) Data type of the audio stream
+        :param channels: (default 1) Number of channels of the audio stream
+        :param rate: (default 16000) Sampling frequency of the audio stream
+        :param chunk: (default 1024) Size of the audio stream buffer
+        :param timeout: (default 2) Time to wait for a new data buffer before stopping recognition due to unavailability
+        of data
+        :param wav_out_fmt: Name format of the recorded audio files
+        """
+        output_dir = os.path.expanduser(output_dir)
+
+        # TODO: Replace this with os.makedirs(output_dir, exist_ok=True) when dropping python2 support
+        makedir_exist_ok(output_dir)
+
+        if " " in wav_out_fmt:
+            raise ValueError("wav_out_fmt cannot have ' ' in the string")
+
+        self.model_dir = model_dir
+        self.model_type = model_type
+        self.output_dir = output_dir
+        self.format = format
+        self.channels = channels
+        self.rate = rate
+        self.chunk = chunk
+        self.timeout = timeout
+        self.wav_out_fmt = wav_out_fmt
+
+        self._p = pyaudio.PyAudio()
```

Comment:
Please a better name than _p. Below I thought it was a file that was being closed.